### DISCUSSION TITLE
Is there a way to limit how many times the bot can check?

### DISCUSSION THREAD

**Discussion by ghost** [jef/streetmerchant#1050] [Help]
I want to use the bot but I am afraid of being banned from newegg and etc from bot detection.
Is there a way to change how many times it will refresh to check?

**Comment by jef** [comment]
Great question! Set `PAGE_SLEEP_MIN` to something like `30000`. That'll scrape a product every 30 seconds.

**Reply by Erik262** [reply]
It doesn't work and checks every link like a machinegun. How to fix that?

**Reply by ashishkjay** [reply]
Sorry not contributing to the issue, but lmao at `checks link like a machinegun`

**Comment by pxpew** [comment]
PAGE_SLEEP_MIN doesn't work on newegg (this is set to 6000)
![bot-ne](https://user-images.githubusercontent.com/75292517/100697423-903eb000-335b-11eb-999d-4657365ebe68.png)

**Comment by MaxCarlson** [comment]
@pxpew from the description of PAGE_SLEEP_MIN 

`Minimum sleep time between queries of the same product page`

It looks like it might not apply to the store as a whole, but for each product page (haven't looked at the code though).

**Comment by ghost** [comment]
I tried different values but nothing seemed to change in how many times it checked

**Comment by dmsolutionz** [comment]
> PAGE_SLEEP_MIN doesn't work on newegg (this is set to 6000)
> ![bot-ne](https://user-images.githubusercontent.com/75292517/100697423-903eb000-335b-11eb-999d-4657365ebe68.png)

PAGE_SLEEP_MIN="6000"
PAGE_SLEEP_MAX="16000"

Make sure you enclose in quotations, could you post your .env file?

**Comment by ghost** [comment]
I have tried this
PAGE_SLEEP_MIN="10000"
PAGE_SLEEP_MAX="20000"

...and it doesn't work. Also I just got banned from Newegg and trying to get an appeal, so that's fun.

Is there no way to have it like "Check website Example1... 10 seconds later... Check website Example 2... etc"
Like in between each search there is a timeout period of how long until it can check the next website

**Comment by Dimitah** [comment]
Would you suggest checking in a larger time window rather than a smaller one to avoid getting banned?

**Comment by tonezz9** [comment]
Are any bans happening on Best Buy? I know they drop a lot less but I'm about to just run it there

**Comment by briankelly** [comment]
@tonezz9 I haven't been banned on best buy, but I did get hit with a 24 hour IP ban on newegg. GL!

**Comment by phreeque** [comment]
I have managed to get banned from Gamestop.com, Not sure how to get that fixed, But I did not let it run for very long
![Capture](https://user-images.githubusercontent.com/9125152/101182068-a1b5df80-361b-11eb-9d37-75db91ab53dd.JPG)
.

**Comment by adrjmz** [comment]
@phreeque try using a vpn when running the script

**Reply by TurkeyPuncher** [reply]
I can't say I'm a VPN expert but I can confirm that this works. Just make sure your VPN service issues you a dynamic IP which should be I think standard for VPN services now. I can see that my B&H requests on my VPN machine has been 24H banned but I can still go to the same page with my second computer. If you've already gotten your current IP banned and you're not willing to wait the 24 hour ban, you should be able to use a VPN to visit the banned pages.

**Comment by phreeque** [comment]
I will moving forward, I am just trying to figure out how to fix this for everyone else in my  house.

**Comment by RavenousSix** [comment]
This setting doesn't work at all. It still checks as fast as it wants.

**Reply by Erik262** [reply]
yes same here, something is broken here..

**Comment by timuriss** [comment]
Agreed that it doesn't work. Have this in my dotenv file and nothing happens:
PAGE_SLEEP_MIN= "10000"
PAGE_SLEEP_MAX= "20000"

**Comment by 0000day** [comment]
What you guys have been looking for is
`PAGE_BACKOFF_MIN`
`PAGE_BACKOFF_MAX`

I know this is an old issue. Just for documentation.

**Reply by diego-coder** [reply]
Why do people keep repeatedly suggesting features that don't apply to this issue?

OP has an issue with the rate on each website. OP does not have an issue with

```
PAGE_BACKOFF_MIN
PAGE_BACKOFF_MAX
PAGE_SLEEP_MIN
PAGE_SLEEP_MAX
```

Those are clearly delineated in the docs and do not address the issue here. And before anyone suggests it, OP does not have an issue with 

`PAGE_TIMEOUT`

The issue, which has been open throughout various threads for months whether in discord or throughout github, is the rate at which the bot repeatedly checks an individual website which results in bans, not the rate that the bot checks an individual product page. 

Let's also reiterate what OP does NOT need here to get this feature: a proxy or VPN.

What is needed is a 

> PAUSE_BETWEEN_CHECKS= 3

in seconds in the env. This is a very simple feature, perhaps one of the first that should have been implemented before anything else. It is a feature that is absent in an otherwise brilliant program, I haven't looked into the code to see where it could be inserted, but I'm sure it could be implemented in a snap.

**Reply by 0000day** [reply]
And I dont still get the difference. If you want a pause between the checks, you are obviously not looking to un-stress your net connection but to limit the amount of requests onto a **page** OR product. Both can be done with the settings stated above. Either limit the **website** with PAGE_BACKOFF_MIN/MAX or limit the product with PAGE_SLEEP_MIN/MAX?

Where is the problem?

**Reply by c0rrre** [reply]
Does PAGE_BACKOFF_MIN/MAX limit the **website** or the **shop**?

Example: When looking for a PS5 and Xbox on Amazon, the bot checks them one after the other and doesn't wait the specified time

**Reply by jef** [reply]
```
PAGE_BACKOFF_MIN
PAGE_BACKOFF_MAX
```

These are related to backing off a _store_ that has given you 403s (by default).

Here is the related documentation in the code: https://github.com/jef/streetmerchant/blob/e2d804027a3ec6a19201ef9ce35dcb4c4dee7025/src/store/model/store.ts#L227-L231

---

```
PAGE_SLEEP_MIN
PAGE_SLEEP_MAX
```

Is related to a _store_ as well, it will determine how long to wait _after_ it's been iterating through the whole store (singular).

---

The idea here is that going to multiple products in a store somewhat common behavior for a normal, manual user. Hence why we are waiting to go through _all_ the links in a store again. As iterating through a whole store constantly could trigger something.

That being said, I think the verbiage is somewhat up to interpretation here, but I also thinking adding in a sleep _per link_ is helpful as @diego-coder said, it's been asked a few times.

**Comment by a2m1** [comment]
As an option to avoid blocking your IP Address you can use tor
Simple setup on Mac
1. Install Tor on Mac using `brew install tor`
2. Run tor with any of the options
  2.1 Start process in command line using`tor` command
  2.2. Start tor as service using `brew service start tor`
  2.3 Previous option with updating the connection `while true; do date; brew services restart tor; sleep 300; done` - it will restart tor service every 300 seconds so you will have a different it
3. Configure the streetmerchant to use tor proxy `socks5://127.0.0.1:9050`

It's not perfect and some addresses might be blocked but still something.

**Reply by jef** [reply]
Nice find.

**Reply by pilosoposerio** [reply]
In Windows (simple steps; non-cli):
1. Download [Tor Browser](https://www.torproject.org/download/) for Windows and install it.
1. Open Tor Browser. Click the `Connect` button.
    1. Auto-connect can be configured on this page: `about:preferences#tor`
    1. Closing the Tor browser will disconnect tor.
1. Edit your `dotenv` and add the following:
    
    ```
    PROXY_ADDRESS=127.0.0.1
    PROXY_PROTOCOL=socks5
    PROXY_PORT=9150
    ```
1. Run.
